File: ch3/clip7/boj1238/main.py
import sys
import heapq

def dijkstra(graph, start, end):
    pq = [(0, start)]
    cost = [float('inf')] * (len(graph))
    cost[start] = 0
    
    while pq:
        current_cost, current_node = heapq.heappop(pq)
        
        if cost[current_node] < current_cost:
            continue
        
        for neighbor, neighbor_cost in graph[current_node]:
            if cost[neighbor] > current_cost + neighbor_cost:
                cost[neighbor] = current_cost + neighbor_cost
                heapq.heappush(pq, (cost[neighbor], neighbor))
    
    return cost

n, m, x = map(int, sys.stdin.readline().split())

forward = [[] for _ in range(n + 1)]
backward = [[] for _ in range(n + 1)]

for _ in range(m):
    s, d, c = map(int, sys.stdin.readline().split())
    forward[s].append((d, c))
    backward[d].append((s, c))

forward_cost = dijkstra(forward, x, 0)
backward_cost = dijkstra(backward, x, 0)

answer = 0

for i in range(1, n + 1):
    answer = max(answer, forward_cost[i] + backward_cost[i])

print(answer)

File: ch3/clip7/boj1238/Main.java
import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt(), m = sc.nextInt(), x = sc.nextInt();
        List<Edge>[] forward = new ArrayList[n + 1];
        List<Edge>[] backward = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) {
            forward[i] = new ArrayList<>();
            backward[i] = new ArrayList<>();
        }
        for (int i = 0; i < m; i++) {
            int s = sc.nextInt(), d = sc.nextInt(), c = sc.nextInt();
            forward[s].add(new Edge(d, c));
            backward[d].add(new Edge(s, c));
        }
        int[] forwardCost = dijkstra(forward, x);
        int[] backwardCost = dijkstra(backward, x);
        int ans = 0;
        for (int i = 1; i <= n; i++) {
            ans = Math.max(ans, forwardCost[i] + backwardCost[i]);
        }
        System.out.println(ans);
    }
    static int[] dijkstra(List<Edge>[] graph, int start) {
        PriorityQueue<Edge> pq = new PriorityQueue<>((o1, o2) -> {
            return o1.cost - o2.cost;
        });
        pq.offer(new Edge(start, 0));
        int[] cost = new int[graph.length];
        final int INF = 1000000000;
        for (int i = 1; i < cost.length; i++) {
            cost[i] = INF;
        }
        cost[start] = 0;
        while (!pq.isEmpty()) {
            Edge now = pq.poll();
            if (cost[now.dist] < now.cost) continue;
            for (Edge next : graph[now.dist]) {
                if (cost[next.dist] > cost[now.dist] + next.cost) {
                    cost[next.dist] = cost[now.dist] + next.cost;
                    pq.offer(new Edge(next.dist, cost[next.dist]));
                }
            }
        }
        return cost;
    }
}

class Edge {
    int dist, cost;

    public Edge(int dist, int cost) {
        this.dist = dist;
        this.cost = cost;
    }
}
